src/InstallerHosts/Elastic.InstallerHosts/Elasticsearch/Tasks/Uninstall/UninstallPluginsTask.cs

```diff
@@ -1,48 +0,0 @@
-﻿using System.Collections.Generic;
-using System.IO.Abstractions;
-using System.Linq;
-using Elastic.Configuration.EnvironmentBased;
-using Elastic.Installer.Domain.Configuration.Wix.Session;
-using Elastic.Installer.Domain.Model.Elasticsearch;
-using Elastic.InstallerHosts.Elasticsearch.Tasks.Install;
-
-namespace Elastic.InstallerHosts.Elasticsearch.Tasks
-{
-	public class UninstallPluginsTask : ElasticsearchInstallationTaskBase
-	{
-		public UninstallPluginsTask(string[] args, ISession session) 
```

Comment:
I know this works today because we have the `PreserveInstallAction` copying folders around but the fact we are sidestepping the `elasticsearch-plugin.bat` does not sit right with me. Although I realize this is now the way to go so no blocker.

We really need to make this an ES core concern e.g `bin/plugin reinstall` could simply update all official plugins currently installed or fail and rollback itself. then if we rollback as a whole on the installer we call `bin/plugin reinstall` itself again.

`</rant>`

Reply:
Agreed; it's a case of trying to fit an existing deployment model into the constraints of the MSI deployment model. The primary reasons that `elasticsearch-plugin.bat` can't be used are that 

1. the flow for updating plugins requires uninstall of current plugin versions and the install of new plugins
2. the plugins directory is not managed by the MSI, so files installed with plugin installation are not associated to components within the installer. Because there is no association, there is no file tracking in case of rollback.

Were `elasticsearch-plugin.bat` able to upgrade existing installed plugins, and rollback in case of error would be great and remove the need to copy the existing plugins directory on start of upgrade, to restore in case of rollback. Even if the responsibility were moved to `elasticsearch-plugin.bat`  however, I can't see how it would accomplish the rollback without caching the existing plugins somewhere, and a mechanism to purge this cache when necessary (since the upgrade process could fail at a later step, after successful plugin update).

An alternative approach would be bundling plugins with the installer as features, but this

1. would increase the size of the MSI database.
2. the database would need to have components for each plugin file and directory in order for tracking to work, which would require it to have knowledge of the deployment destination of each.
3. wouldn't help third party plugins, and what happens to those on upgrade.